src/PetriEngine/Colored/Reduction/RedRulePreAgglomeration.cpp

```diff
@@ -0,0 +1,213 @@
+            const auto postsize = place._post.size();
+            bool ok = true;
+            uint32_t i = 0, j = 0;
+            while (i < presize && j < postsize) {
+                if (place._pre[i] < place._post[j])
+                    i++;
+                else if (place._pre[j] < place._post[i])
+                    j++;
+                else {
+                    ok = false;
+                    break;
+                }
+            }
+
+            if (!ok) continue;
+
+            // S2
+            std::vector<bool> todo (postsize, true);
+            bool todoAllGood = true;
+            // S10-11; Do we need to check?
+            std::vector<bool> kIsAlwaysOne (postsize, true);
+            // Visitor for translating ArcExpressions into the info we need.
+            ArcExpressionVisitor arvis = ArcExpressionVisitor();
+
+            for (const auto& prod : place._pre){
+                const Transition& producer = red.transitions()[prod];
+                // S4, S6.2
+                if(producer.inhibited || producer.input_arcs.size() != 1){
+                    ok = false;
+                    break;
+                }
+
+                // Have arvis visit the arc from producer to place
+                arvis.reset();
+                red.getOutArc(producer, pid)->expr->visit(arvis);
+                uint32_t kw = 1;
+
+                if(arvis.ok() && arvis.singleColor()){
+                    kw = arvis.colorblindWeight();
+                } else {
+                    ok = false;
+                    break;
+                }
+
+                // S1, S9
+                for (uint32_t n = 0; n < place._post.size(); n++) {
+                    arvis.reset();
+                    red.getInArc(pid, red.transitions()[place._post[n]])->expr->visit(arvis);
+                    uint32_t w = arvis.colorblindWeight();
+                    // S1, S9
+                    if (!arvis.ok() || !arvis.singleColor() || kw % w != 0) {
+                        todo[n] = false;
+                        todoAllGood = false;
+                        continue;
+                    } else if (kw != w) {
+                        kIsAlwaysOne[n] = false;
+                    }
+                }
+
+                // Check if we have any qualifying consumers left
+                if (!todoAllGood && std::lower_bound(todo.begin(), todo.end(), true) == todo.end()){
+                    ok = false;
+                    break;
+                }
+
+                for (const auto& prearc : producer.input_arcs){
+                    const Place& preplace = red.places()[prearc.place];
+                    // S6.3, S5.2
+                    if (preplace.inhibitor || inQuery[prearc.place] > 0){
+                        ok = false;
+                        break;
+                    } else if (queryType != Reach) {
+                        // For reachability, we can do free agglomeration which avoids this condition
```

Comment:
I følge definitionen er disse conditions afhængig af deadlocks i query'en og ikke om det er reachability (de to er ikke modsætninger). Så condition her burde være ` if (preserveLoops) `